src/Skynet.Protocol/ChannelMessage.cs

```diff
@@ -0,0 +1,151 @@
+
+            byte[] hmacKey = key.Slice(0, 32).ToArray();
+            byte[] aesKey = key.Slice(32, 32).ToArray();
+            ReadContent(AesStatic.EncryptWithHmac(PacketContent.Value, hmacKey, aesKey));
+            Array.Clear(hmacKey, 0, 32);
+            Array.Clear(aesKey, 0, 32);
+        }
+
+        public void Encrypt(ReadOnlySpan<byte> key)
+        {
+            if (key.Length != 64) throw new ArgumentOutOfRangeException(nameof(key), key.Length, "The key must be exactly 64 bytes long.");
+
+            byte[] hmacKey = key.Slice(0, 32).ToArray();
+            byte[] aesKey = key.Slice(32, 32).ToArray();
+            PacketContent = AesStatic.EncryptWithHmac(WriteContent(), hmacKey, aesKey);
+            Array.Clear(hmacKey, 0, 32);
+            Array.Clear(aesKey, 0, 32);
+        }
+
+        protected sealed override void ReadPacketInternal(PacketBuffer buffer, PacketRole role)
+        {
+            PacketVersion = buffer.ReadByte();
+            ChannelId = buffer.ReadInt64();
+            if (role == PacketRole.Client)
+                SenderId = buffer.ReadInt64();
+            MessageId = buffer.ReadInt64();
+            if (role == PacketRole.Client)
+            {
+                SkipCount = buffer.ReadInt64();
+                DispatchTime = buffer.ReadDateTime();
+            }
+            MessageFlags = (MessageFlags)buffer.ReadByte();
+            if (MessageFlags.HasFlag(MessageFlags.ExternalFile))
+                FileId = buffer.ReadInt64();
+
+            PacketContent = buffer.ReadMediumByteArray();
+
+            if (MessageFlags.HasFlag(MessageFlags.Unencrypted))
+            {
+                ReadContent(PacketContent.Value);
+            }
+
+            int length = buffer.ReadUInt16();
+            for (int i = 0; i < length; i++)
+            {
+                Dependencies.Add(new Dependency(buffer.ReadInt64(), buffer.ReadInt64()));
+            }
+        }
+
+        protected sealed override void WritePacketInternal(PacketBuffer buffer, PacketRole role)
+        {
+            buffer.WriteByte(PacketVersion);
+            buffer.WriteInt64(ChannelId);
+            if (role == PacketRole.Server)
+                buffer.WriteInt64(SenderId);
+            buffer.WriteInt64(MessageId);
+            if (role == PacketRole.Server)
+            {
+                buffer.WriteInt64(SkipCount);
+                buffer.WriteDateTime(DispatchTime);
+            }
+            buffer.WriteByte((byte)MessageFlags);
+            if (MessageFlags.HasFlag(MessageFlags.ExternalFile))
+                buffer.WriteInt64(FileId);
+
+            if (PacketContent == null)
+            {
+                if (MessageFlags.HasFlag(MessageFlags.Unencrypted))
+                    PacketContent = WriteContent();
+                else
+                    throw new InvalidOperationException("A message without MessageFlags.Unencrypted has be encrypted before it can be written.");
```

Comment:
```suggestion
                    throw new InvalidOperationException("A message without MessageFlags.Unencrypted has to be encrypted before it can be written.");
```
Also, what exactly does this error message mean?